File: music_app/services/navigation_tree.py
"""Server renderer for the shared flat NavigationTreeItem template."""
from pathlib import Path
import re

from markupsafe import Markup, escape

_ITEM_TEMPLATE = (Path(__file__).resolve().parent.parent / "templates" / "components" / "navigation-tree-item.html").read_text(encoding="utf-8")


def render_navigation_tree_item(*, label, href="#", key="", selected=False, icon="", count=None, variant="artists", attributes=None, action=False) -> Markup:
    settings = variant == "settings"
    legacy = "settings-nav-item" if settings else "artist-link"
    active_class = " is-active" if settings else " active"
    tag = "button" if action else "a"
    values = {
        "class": legacy + (active_class if selected else "") + " navigation-tree-item" + (" is-selected" if selected else ""),
        "data-navigation-tree-item": "settings" if settings else "artists",
        "data-navigation-tree-key": key,
    }
    if action:
        values["type"] = "submit"
    else:
        values["href"] = href if re.match(r"^(?:/(?!/)|#)", str(href)) else "#"
    for name, value in (attributes or {}).items():
        if re.fullmatch(r"data-[a-z0-9-]+", name) and not name.startswith("data-navigation-tree-"):
            values[name] = value
    if selected:
        values["aria-current"] = "page" if settings else "true"
    attrs = " ".join(f'{name}="{escape(value)}"' for name, value in values.items())
    icon_html = f'<span class="navigation-tree-icon" aria-hidden="true">{escape(icon)}</span>' if icon else ""
    count_html = "" if count is None else f'<span class="navigation-tree-count artist-count">{escape(count)}</span>'
    return Markup(_ITEM_TEMPLATE % (tag, attrs, icon_html, escape(label), count_html, tag))

File: music_app/static/js/navigation-tree.js
/* Flat navigation presentation shared by Artists and Settings. */
(() => {
  'use strict';
  if (window.NavigationTree) return;
  const escape = value => String(value ?? '').replace(/[&<>"']/g, char => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[char]));
  // HTML slots contain component-rendered markup; resource text uses escaped fields.
  function renderItem({ label, href = '#', key = '', selected = false, icon = '', count = null, variant = 'artists', attributes = {}, action = false, artworkHtml = '', subtitle = '', year = '', countHidden = false, trailingHtml = '', draggable = false, className = '' } = {}) {
    const template = document.getElementById('navigation-tree-item-template')?.textContent;
    if (!template) throw new Error('NavigationTreeItem template is missing.');
    const settings = variant === 'settings';
    const wide = variant === 'wide';
    const panel = variant === 'panel' || wide;
    const legacy = settings ? 'settings-nav-item' : 'artist-link';
    const tag = action ? 'button' : 'a';
    const values = {
      class: legacy + (selected ? (settings ? ' is-active' : ' active') : '') + ' navigation-tree-item' + (wide ? ' navigation-tree-item--wide utility-list-item' : '') + (selected ? ' is-selected' + (wide ? ' is-active' : '') : '') + (className ? ' ' + className : ''),
      'data-navigation-tree-item': wide ? 'wide' : panel ? 'panel' : (settings ? 'settings' : 'artists'),
      'data-navigation-tree-key': key,
    };
    if (action) values.type = panel ? 'button' : 'submit';
    else values.href = /^(?:\/(?!\/)|#)/.test(String(href)) ? href : '#';
    if (draggable) values.draggable = 'true';
    for (const [name, value] of Object.entries(attributes)) {
      if (/^data-[a-z0-9-]+$/.test(name) && !name.startsWith('data-navigation-tree-')) values[name] = value;
    }
    if (selected) values['aria-current'] = settings ? 'page' : 'true';
    const attrs = Object.entries(values).map(([name, value]) => name + '="' + escape(value) + '"').join(' ');
    const identity = wide || (panel && (subtitle || year)) ? '<span class="utility-list-item-title">' + escape(label) + '</span><span class="utility-list-item-meta">' + escape(subtitle) + (year ? (subtitle ? ' · ' : '') + escape(year) : '') + '</span>' : escape(label);
    const slots = [tag, attrs, wide ? '<span class="navigation-tree-artwork">' + artworkHtml + '</span>' : icon ? '<span class="navigation-tree-icon" aria-hidden="true">' + escape(icon) + '</span>' : '', identity,
      (count === null ? '' : '<span class="navigation-tree-count artist-count"' + (countHidden ? ' hidden' : '') + '>' + escape(count) + '</span>') + (wide ? trailingHtml : ''), tag];
    let index = 0;
    return template.replace(/%s/g, () => slots[index++]);
  }
  function setItemSelected(item, selected) {
    if (!item?.classList) return;
    const settings = item.getAttribute('data-navigation-tree-item') === 'settings' || item.classList.contains('settings-nav-item');
    const wide = item.getAttribute('data-navigation-tree-item') === 'wide';
    item.classList.toggle(settings ? 'is-active' : 'active', Boolean(selected));
    if (wide) item.classList.toggle('is-active', Boolean(selected));
    item.classList.toggle('is-selected', Boolean(selected));
    if (selected) item.setAttribute('aria-current', settings ? 'page' : 'true');
    else item.removeAttribute('aria-current');
  }
  function updateItem(item, { label = '', subtitle = '', year = '', count = null, artworkHtml, artworkLabel } = {}) {
    for (const [selector, value] of [
      ['.utility-list-item-title', label],
      ['.utility-list-item-meta', subtitle + (year ? ' · ' + year : '')],
      ['.navigation-tree-count', count === null ? '' : count],
    ]) {
      const field = item.querySelector?.(selector);
      if (field && field.textContent !== String(value)) field.textContent = String(value);
    }
    const artwork = item.querySelector?.('.navigation-tree-artwork');
    if (artwork && artworkHtml !== undefined) artwork.innerHTML = artworkHtml;
    if (artwork && artworkLabel !== undefined) {
      artwork.querySelector?.('img')?.setAttribute?.('alt', artworkLabel);
      artwork.querySelector?.('.album-artbox')?.setAttribute?.('aria-label', artworkLabel);
    }
  }
  function setSelection(root, key) {
    root?.querySelectorAll('[data-navigation-tree-item]').forEach(item => {
      const selected = String(item.getAttribute('data-navigation-tree-key') || '') === String(key);
      if (item.classList.contains('is-selected') !== selected) setItemSelected(item, selected);
    });
  }
  window.NavigationTree = { renderItem, renderItems: items => items.map(renderItem).join(''), updateItem, setItemSelected, setSelection };
})();
